### DISCUSSION TITLE
Upgrading Aurora MySQL with minimal downtime

### DISCUSSION THREAD

**Discussion by erictompkins** [gruntwork-io/knowledge-base#502] [Help] [answered]
We currently have an Aurora RDS cluster with read and write instances. ([r:terraform-aws-data-storage](https://github.com/gruntwork-io/terraform-aws-data-storage))

<img width="983" alt="image" src="https://user-images.githubusercontent.com/4851312/178852113-6bd99370-dc0f-4b95-b07f-1e0fac4cff21.png">

We need to upgrade the Aurora version and also the instance size. My understanding is that both of those updates require downtime. Is that correct?

Currently, the engine version is `5.6.mysql_aurora.1.22.2`.  We do have Multi-AZ enabled with 2 zones.

![CleanShot 2022-07-13 at 19 10 46@2x](https://user-images.githubusercontent.com/4851312/178852325-6bfbdb9b-9e8e-418c-8ff9-cbd47de943f1.png)

Do you have any recommendations to do an upgrade with minimal downtime?

I've done some research and doing a blue/green deployment may be a good way, but I'm not sure if that's the best way and how to set that up with Terraform.

Dev ops is not my primary function (we're a very small team) so any recommendations are welcomed. Thank you.

---

<ins datetime="2022-07-13T23:11:49Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108982">Tracked in ticket #108982</a></p>
</ins>

**Comment by rhoboat** [comment, marked as answer]
Note, whatever you do, always schedule a maintenance window for this work, and let everyone know about the window. Minimal downtime is still downtime, and things can always go wrong.

Some important resources:

- https://aws.amazon.com/blogs/database/performing-major-version-upgrades-for-amazon-aurora-mysql-with-minimum-downtime/
- https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.Upgrading.html
- https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Updates.MajorVersionUpgrade.html#AuroraMySQL.Upgrading.Compatibility

At `xlarge`, you're looking at several hours of downtime.

Whether you do this update in place, or a blue/green deployment, the in-place and the green deployment can be upgraded the same way: 
- Change the `engine_version` number in your terragrunt/terraform configuration and run `terragrunt apply` or `terraform apply`. 
- For the instance type, change the `instance_type` or `instance_class` configuration (our `rds` module names this variable `instance_type`, but the resource calls it `instance_class` internally). You can do these both in the same changeset for the a single `apply`.
- Some links to our module code: [aurora](https://github.com/gruntwork-io/terraform-aws-data-storage/blob/master/modules/aurora/main.tf#L298-L301) and [rds](https://github.com/gruntwork-io/terraform-aws-data-storage/blob/master/modules/rds/main.tf#L141-L144)

If you can afford it, blue/green deployment is the better choice. You should still plan for the minimal downtime which exists in blue/green deployment, during the switch over. Note that a lot of resources out there say that updating just the instance type (if you wanted to do the version and instance type as separate upgrade steps) is a zero downtime upgrade on Aurora RDS with multi-AZ setup. But this is not true. People still report downtime of 10 - 20 minutes. So my take is that if you're going to update both version and instance type, might as well do them together.

Setting up the green environment replica can be done without using ClickOps as well. If you're familiar enough with the module code that currently manages your RDS cluster, you'd make a copy of that, put it in a new folder. If you're using our `infrastructure-live` folder structure, you would choose which account/region/environment it belongs in. If you're using our Gruntwork Pipelines (which comes with the Reference Architecture) you can push this new folder to your repo, have your CI pick up the change and run apply for you (with your approval). But you can instead opt to run `apply` locally, authing from your machine to AWS, from that folder. This might lock up your machine, which can be a downside. The plan should show what it will create. Once it's up, update the configuration and apply it. Then wait for the instances to be available. _This is where I'm not sure--your local machine might be locked up running this apply process for the full duration of the upgrade!_ When you're ready to destroy, run `destroy` manually from the original folder where the blue environment configuration is.

**Reply by erictompkins** [reply]
Thank you, that is helpful.

As part of the blue/green deployment we would need to sync the database data. Is that something that can be done with Terraform or do we need to do it another way?

**Reply by rhoboat** [reply]
Yes, great question, key point that's missing from my response. To give you a perfect response, we're going to have to do some spiking (running through a blue/green upgrade). Both myself and my colleague are interested in learning how this process actually works in Terraform. 

Here are my some raw thoughts until we can provide a definitive answer:

You'll need to replicate the data at the time right before the upgrade. After the upgrade is complete and the green environment instances are available, you'll need to replicate again to pick up the new changes.

Clone option
------------
The [first link above](https://aws.amazon.com/blogs/database/performing-major-version-upgrades-for-amazon-aurora-mysql-with-minimum-downtime/) is a complete walkthrough of the upgrade process, done without Terraform. They recommend doing the first replication via cloning, and the second one via binary log replication. There is support in our modules for cloning, with [`restore_source_cluster_identifier`](https://github.com/gruntwork-io/terraform-aws-data-storage/blob/master/modules/aurora/variables.tf#L202). This leverages the underlying functionality described in HashiCorp's AWS provider docs [here](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/rds_cluster#restore_to_point_in_time-argument-reference), along with how to use it. To reference the existing cluster in as the restore_source_cluster_identifier, you'll need to actually modify the existing configuration, adding a new module block for the clone, and that way you can reference the original by its module address (`module.my_primary_aurora.cluster_arn`). However, then to truly be a blue/green deployment, that green cluster would need to move out into its own environment, which would mess up Terraform state. There's no good way forward that I can see once you stop managing the resources with Terraform.

For binary log replication, it looks like the only way is manually within MySQL itself (Click Ops again).

Replica option
--------

Another thought is to use a replica setup. The replication can be set up in the module configuration. Rather than copy the existing configuration, instead modify the existing configuration with a new module block for the replica. Then, if you were not using Terraform, you'd want to promote the replica to a separate green environment. However the same issue applies--how to promote the cluster if you're using Terraform.

[Example of rds-mysql-with-cross-region-replica](https://github.com/gruntwork-io/terraform-aws-data-storage/blob/6722489760579a9a93e86838caada3fee99fe6c5/examples/rds-mysql-with-cross-region-replica/main.tf). 
[Example of aurora-with-cross-region-replica](https://github.com/gruntwork-io/terraform-aws-data-storage/blob/6722489760579a9a93e86838caada3fee99fe6c5/examples/aurora-with-cross-region-replica/main.tf). 
Please note that if you're using AWS Provider v4, when you set `replicate_source_db` for the rds example, you must unset (remove) `engine` and `engine_version`. (In the aurora module, the same variable is called `replication_source_identifier`.)

Snapshot option
-------

This is another one we should look into. It might be easiest to create a snapshot using our [various Lambda snapshot modules](https://github.com/gruntwork-io/terraform-aws-data-storage/tree/2c25ca62b30f5642af7f7dcd462645147a00f456/modules).

It's likely that the practical solution is a combo of Terraform and Click Ops :disappointed:.

**Reply by erictompkins** [reply]
All of those options make sense, but like you, we're at a loss of how to do this all with Terraform.

We are hoping to do this upgrade within the next few weeks. Do you think that you would have time to figure out how to do this with Terraform in that time frame?

If a combo of Terraform form and the AWS console is the only option, would it be possible to import the new Aurora cluster and its configuration into Terraform, or would that be a major pain? If it's possible, do you have recommendations or examples of doing that?

**Reply by yorinasub17** [reply]
Hi @erictompkins , I spent some time looking into this, and have tried the following approach which seems to work to implement the blue-green deployment:

NOTE: in the document below, blog post refers to [this AWS Blog Post](https://aws.amazon.com/blogs/database/performing-major-version-upgrades-for-amazon-aurora-mysql-with-minimum-downtime/).

1. If you haven't already, create a DNS entry for your database. This will make it easier to switch over between blue and green without updating your app. Make sure all your apps are using the known stable DNS endpoint to access the DB, rather than the raw RDS endpoint.

1. Set up binlog format to be `MIXED` as described in the blog post.
    1. Configure the parameter group to setup binlog format. If you are using the service catalog, you can set this using [db_cluster_custom_parameter_group](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/modules/data-stores/aurora/variables.tf#L95https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/modules/data-stores/aurora/variables.tf#L95). If not, you can use the `aws_rds_cluster_parameter_group` terraform resource in your `infrastructure-modules` repo.
    2. Follow the blog post recommendations to restart the writer instance (NOTE: requires downtime).
    3. Follow the blog post commands to validate the configuration and update the bin log retention period. Since I expect the following steps to take a long time, I recommend setting the retention period to 3 days or longer (set to 72).

1. Make sure a snapshot exists that is after you had made the changes above (use the AWS Console to check snapshots). You want the new green environment database to be restored to a point after the database started recording the binary logs, so that it can replay the logs to catch up, and stay up to date.

1. Create a fast clone RDS cluster with the same settings as the original (green environment).
    1. Do this by copy pasting your `terragrunt.hcl` configuration for the original db.
    2. Update the `name` input variable so that it won't collide with the original.
    5. Set the `restore_source_cluster_identifier` input variable to the RDS cluster name of the original db. This will restore the DB to the latest restorable time available.
    6. Set the instance type to be the new type that you want to use.
    7. Apply the new config to create the green environment with a point in time restore.

1. Once the new database (green) comes up, upgrade the database engine version of the green environment using the web console (`terraform` currently does not support updating from aurora v1 to v2). Make sure to apply immediately.
1. Wait for the upgrade to finish. **Note that this may take a long time**. The key thing to be careful here is to make sure that this step doesn't take longer than the binlog retention period you set above. If it does, then you may have data loss and you will need to start over with a longer binlog retention period.
1. Update the Terraform code to reflect the new version (`engine` or `engine_version`) so that it is in sync. Note that the parameter group may need to be recreated. During the update, AWS will automatically associate the database with the default parameter group for the new MySQL version.
1. Once the green environment is updated to look like how you want, hook up the green environment to be a streaming replica of the blue environment. **This will have to be done manually using the AWS Console and a SQL client**. Refer to the blog post for information on how to do this (section "Setting up binary log replication").
    - NOTE: When setting up the replication user on the blue environment, it is best to do `CREATE USER repl_user
      IDENTIFIED by '<PASSWORD>';` and `GRANT REPLICATION SLAVE ON *.* TO 'repl_user';`

1. Wait for the green environment to catch up, and is in sync. You can test this by writing data to a table in the main blue database, and observing for that test data in the green environment.
1. Once the database is reasonably in sync, start the procedure to fail over to the new environment. Refer to the blog post for the safest approach for doing this. **NOTE: you will have downtime for write access during this time.**
1. Once the new database is ready, point the DNS entry for your database to the green environment. Wait for the DNS entry to propagate. **NOTE: You will have downtime for write access during this time.**
1. Once the DNS entry propagates, your app should be able to start making writes again, with all writes being routed to
the green environment. Do a smoke test through your app functionality, and make sure that the data is up to date.
1. Once you confirm the environment is good, you can tear down the blue environment to complete the procedure. This can be done by running a `terragrunt destroy`.
1. Finally, double check to make sure all the Terraform inputs for the green environment reflect the state of AWS. You can do this by iterating between `plan` and configuration updates until Terraform stops detecting a drift.
1. (**Optional, but recommended**) Update the bin log retention period to be a more reasonable period (e.g., 12 hours) to avoid filling up your disk space.

I recommend doing a dry run in a completely new dev cluster first to familiarize yourself with the above procedure in an environment that you can easily teardown and restore prior to attempting this with any live environment.

Let me know if you have any questions regarding the above!

**Reply by erictompkins** [reply]
Thank you very much for that detailed plan. We will give this a try soon and let you know if we have any further questions or need any other help.

**Reply by erictompkins** [reply]
What is the highest version of Aurora MySQL that Gruntworks supports? Is it 2.3 or does it support 3.0?
It looks like 3.02.0 is the last version supported by AWS.

**Reply by yorinasub17** [reply]
Gruntwork supports version 3, but note that you can't upgrade straight to version 3 from version 1. You must upgrade to version 2 first.

**Reply by erictompkins** [reply]
Ok - thank you. I figured that the upgrade would have to be incremental. 

The documentation for the `engine` parameters in the `aurora` module says that `aurora-mysql` is for MySQL 5.7. It looks like Aurora version 3 is for MySQL 8.0. I just wanted to confirm that `aurora-mysql` would work with Aurora 3 or if there is another parameter to change.  The `engine_version` parameter would also updated to the correct version.

**Reply by yorinasub17** [reply]
Yup `aurora-mysql` supports both v2 and v3, and you just have to set the `engine_version` parameter to the desired version (e.g., `8.0.mysql_aurora.3.02.0`). Note that if you do it through the console, `terraform` will self update to the version you want.
